src/serialize.ts

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import utils = require('./utils');
+import kernel = require('./kernel');
+
+import IKernelMsg = kernel.IKernelMsg;
+
+
+/**
+ * Deserialize and return the unpacked message.
+ */
+export
+function deserialize(data: ArrayBuffer | string): IKernelMsg {
+  var value: IKernelMsg;
+  if (typeof data === "string") {
+    value = JSON.parse(data);
+  } else {
+    value = deserializeBinary(data);
+  }
+  return value;
+}
+
+
+/**
+ * Serialize a kernel message for transport.
+ */
+export
+function serialize(msg: IKernelMsg): string | ArrayBuffer {
+  var value: string | ArrayBuffer;
+  if (msg.buffers && msg.buffers.length) {
+    value = serializeBinary(msg);
+  } else {
+    value = JSON.stringify(msg);
+  }
+  return value;
+}
+
+
+/**
+ * Deserialize a binary message to a Kernel Message.
+ */
+function deserializeBinary(buf: ArrayBuffer): IKernelMsg {
+  var data = new DataView(buf);
+  // read the header: 1 + nbufs 32b integers
+  var nbufs = data.getUint32(0);
+  var offsets: number[] = [];
+  for (var i = 1; i <= nbufs; i++) {
+    offsets.push(data.getUint32(i * 4));
+  }
+  var json_bytes = new Uint8Array(buf.slice(offsets[0], offsets[1]));
```

Comment:
what happens if `offsets.length < 2`?

Maybe accept as-is for now, but open an issue which documents the need to address it.